### PR TITLE
[#IOPID-384] notice login email durable function

### PR DESCRIPTION
**Pull request by arcogabbo**
This PR depends on #281
This PR depends on https://github.com/pagopa/io-messages-email-templates/pull/13
This PR depends on https://github.com/pagopa/io-infra/pull/614

#### List of Changes

* added 3 activities: `SendTemplatedLoginEmailActivity`,`GetMagicCodeActivity`,`GetGeoLocationDataActivity`
* added 1 orchestrator: `NoticeLoginEmailOrchestrator`
* added 1 trigger: `NoticeLoginEmailTrigger`

#### Motivation and Context

sending an email whenever a user completes a full login flow

#### How Has This Been Tested?

unit tests, integration tests via io-mock project

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Review by gquadrati — CHANGES_REQUESTED** [review]

File: NoticeLoginEmailTrigger/__tests__/handler.test.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  it("should return internal error when the orchestrator fails to start", async () => {
+    jest
+      .spyOn(durableUtils, "startOrchestrator")
+      .mockReturnValueOnce(TE.left(new Error("error")));
+    const result = await NoticeLoginEmailHandler()(
+      context as any,
+      aValidTriggerPayload
+    );
+    const res = MockResponse();
+    result.apply(res);
```
[the remark this answers is not in the document]
Comment (arcogabbo) [reply NoticeLoginEmailTrigger/__tests__/handler.test.ts]: nice catch. solved in https://github.com/pagopa/io-functions-app/pull/283/commits/5d3bbbbe32125b4f85d59f551ce72f2aff3b68d5

File: SendTemplatedLoginEmailActivity/handler.ts [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        }),
+        TE.map(result => {
+          const messageInfo = result.value;
+
+          // on success, track a custom event with properties of transport used
+          // see https://github.com/pagopa/io-functions-commons/blob/master/src/utils/nodemailer.ts
+          // note: the extra properties will be defined only when using a MultiTransport
+          createTracker(ai.defaultClient).profile.traceEmailValidationSend(
+            typeof messageInfo === "object" ? messageInfo : {}
+          );
```
[the remark this answers is not in the document]
Comment (arcogabbo) [reply SendTemplatedLoginEmailActivity/handler.ts]: done in https://github.com/pagopa/io-functions-app/pull/283/commits/5d3bbbbe32125b4f85d59f551ce72f2aff3b68d5

File: SendTemplatedLoginEmailActivity/index.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
+import { getMailerTransporter } from "@pagopa/io-functions-commons/dist/src/mailer";
+import { initTelemetryClient } from "../utils/appinsights";
+import { getConfigOrThrow } from "../utils/config";
+import { getSendLoginEmailActivityHandler } from "./handler";
+
+const config = getConfigOrThrow();
+
+// Email data
+const EMAIL_TITLE = "LOGIN EMAIL";
```
[the remark this answers is not in the document]
Comment (arcogabbo) [reply SendTemplatedLoginEmailActivity/index.ts]: done in https://github.com/pagopa/io-functions-app/pull/283/commits/5d3bbbbe32125b4f85d59f551ce72f2aff3b68d5

File: GetGeoLocationDataActivity/handler.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            geoLocationService.getGeoLocationForIp(activityInput.ip_address),
+          E.toError
+        ),
+        TE.mapLeft(_error =>
+          ActivityResultFailure.encode({
+            kind: "API_CALL_FAILURE",
+            reason: "call not yet implemented"
+          })
+        )
+      )
```
[the remark this answers is not in the document]
Comment (arcogabbo) [reply GetGeoLocationDataActivity/handler.ts]: done in https://github.com/pagopa/io-functions-app/pull/283/commits/5d3bbbbe32125b4f85d59f551ce72f2aff3b68d5

File: NoticeLoginEmailOrchestrator/handler.ts [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    ip_address,
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    identity_provider
+  } = orchestratorInput;
+
+  // Log the input
+  context.log.verbose(
+    `${logPrefix}|INPUT=${JSON.stringify(orchestratorInput)}`
+  );
```
[the remark this answers is not in the document]
Comment (arcogabbo) [reply NoticeLoginEmailOrchestrator/handler.ts]: I hashed all the values you mentioned plus the email in https://github.com/pagopa/io-functions-app/pull/283/commits/5d3bbbbe32125b4f85d59f551ce72f2aff3b68d5. thank you for the recommendation.

File: NoticeLoginEmailOrchestrator/handler.ts [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+    email,
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    date_time,
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    device_name,
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    ip_address,
+    // eslint-disable-next-line @typescript-eslint/naming-convention
+    identity_provider
+  } = orchestratorInput;
```
[the remark this answers is not in the document]
Comment (arcogabbo) [reply NoticeLoginEmailOrchestrator/handler.ts]: done in https://github.com/pagopa/io-functions-app/pull/283/commits/5d3bbbbe32125b4f85d59f551ce72f2aff3b68d5

File: NoticeLoginEmailTrigger/__tests__/handler.test.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+    expect(result.kind).toBe("IResponseErrorInternal");
```
[the remark this answers is not in the document]
Comment (arcogabbo) [reply NoticeLoginEmailTrigger/__tests__/handler.test.ts]: done in https://github.com/pagopa/io-functions-app/pull/283/commits/5d3bbbbe32125b4f85d59f551ce72f2aff3b68d5

File: NoticeLoginEmailTrigger/index.ts
Comment (BurnedMarshal) [inline NoticeLoginEmailTrigger/index.ts:1]: IMHO the folder name should be without the `Trigger` suffix. (es. `NoticeLoginEmail`)

File: GetGeoLocationDataActivity/handler.ts [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+      });
+    }),
+    TE.fromEither,
+    // TODO: implement the actual call to geo location service
+    TE.chain(activityInput =>
+      pipe(
+        TE.tryCatch(
+          () =>
+            geoLocationService.getGeoLocationForIp(activityInput.ip_address),
+          E.toError
```
[the remark this answers is not in the document]
Comment (arcogabbo) [reply GetGeoLocationDataActivity/handler.ts]: done in https://github.com/pagopa/io-functions-app/pull/283/commits/40abd733fc1a1797b76bebbfdca9a860f6c0928d

File: NoticeLoginEmailOrchestrator/handler.ts [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+    if (E.isLeft(errorOrGeoLocationServiceResponse)) {
+      // we let geo_location be undefined.
+      // the SendTemplatedLoginEmailActivity will decide what email template to use based on the geo_location value
+      if (!TransientApiCallFailure.is(geoLocationActivityResult)) {
+        throw OrchestratorFailureResult.encode({
+          kind: "FAILURE",
+          reason: readableReportSimplified(
+            errorOrGeoLocationServiceResponse.left
+          )
+        });
```
[the remark this answers is not in the document]
Comment (arcogabbo) [reply NoticeLoginEmailOrchestrator/handler.ts]: done in https://github.com/pagopa/io-functions-app/pull/283/commits/40abd733fc1a1797b76bebbfdca9a860f6c0928d

File: NoticeLoginEmailOrchestrator/handler.ts [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+      // the SendTemplatedLoginEmailActivity will decide what email template to use based on the magic_code value
+      if (!TransientApiCallFailure.is(magicCodeActivityResult)) {
+        throw OrchestratorFailureResult.encode({
+          kind: "FAILURE",
+          reason: readableReportSimplified(errorOrMagicLinkServiceResponse.left)
+        });
+      }
+    } else {
+      magic_code = errorOrMagicLinkServiceResponse.right.value.magic_code;
+    }
```
[the remark this answers is not in the document]
Comment (arcogabbo) [reply NoticeLoginEmailOrchestrator/handler.ts]: done in https://github.com/pagopa/io-functions-app/pull/283/commits/40abd733fc1a1797b76bebbfdca9a860f6c0928d

File: GetMagicCodeActivity/handler.ts [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+          E.toError
+        ),
+        TE.mapLeft(_ =>
+          ActivityResultFailure.encode({
+            kind: "API_CALL_FAILURE",
+            reason: "call not yet implemented"
+          })
+        )
+      )
+    ),
```
[the remark this answers is not in the document]
Comment (arcogabbo) [reply GetMagicCodeActivity/handler.ts]: done in https://github.com/pagopa/io-functions-app/pull/283/commits/40abd733fc1a1797b76bebbfdca9a860f6c0928d

File: NoticeLoginEmailTrigger/index.ts
Comment (arcogabbo) [reply NoticeLoginEmailTrigger/index.ts:1]: done in https://github.com/pagopa/io-functions-app/pull/283/commits/067b2282ce86b9aeab060de9a9911bda60f6ece6

**Comment by arcogabbo** [comment]
> SonarCloud Quality Gate failed.    [Quality Gate failed](https://sonarcloud.io/dashboard?id=pagopa_io-functions-app&pullRequest=283)
> 
> [Bug](https://sonarcloud.io/project/issues?id=pagopa_io-functions-app&pullRequest=283&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=pagopa_io-functions-app&pullRequest=283&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=pagopa_io-functions-app&pullRequest=283&resolved=false&types=BUG) [Vulnerability](https://sonarcloud.io/project/issues?id=pagopa_io-functions-app&pullRequest=283&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=pagopa_io-functions-app&pullRequest=283&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=pagopa_io-functions-app&pullRequest=283&resolved=false&types=VULNERABILITY) [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=pagopa_io-functions-app&pullRequest=283&resolved=false&types=SECURITY_HOTSPOT) [A](https://sonarcloud.io/project/security_hotspots?id=pagopa_io-functions-app&pullRequest=283&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=pagopa_io-functions-app&pullRequest=283&resolved=false&types=SECURITY_HOTSPOT) [Code Smell](https://sonarcloud.io/project/issues?id=pagopa_io-functions-app&pullRequest=283&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=pagopa_io-functions-app&pullRequest=283&resolved=false&types=CODE_SMELL) [0 Code Smells](https://sonarcloud.io/project/issues?id=pagopa_io-functions-app&pullRequest=283&resolved=false&types=CODE_SMELL)
> 
> [No Coverage information](https://sonarcloud.io/component_measures?id=pagopa_io-functions-app&pullRequest=283) No Coverage information [](https://sonarcloud.io/component_measures?id=pagopa_io-functions-app&pullRequest=283&metric=new_duplicated_lines_density&view=list) [5.6% Duplication](https://sonarcloud.io/component_measures?id=pagopa_io-functions-app&pullRequest=283&metric=new_duplicated_lines_density&view=list)
> 
> idea Catch issues before they fail your Quality Gate with our IDE extension sonarlint [SonarLint](https://www.sonarsource.com/products/sonarlint/features/connected-mode/?referrer=sonarcloud-welcome)

this duplication error will be removed after we implement the actual logic and the actual service clients

**Review by gquadrati — APPROVED** [review]
LGTM, good work! 🚀  Thanks @arcogabbo

**Pull request merged**